daemon/state.go

```diff
@@ -97,21 +97,21 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
 	endpointmanager.Mutex.Lock()
 	nEndpoints := len(endpointmanager.Endpoints)
 	wg := make(chan bool, nEndpoints)
-	for k := range endpointmanager.Endpoints {
-		ep := endpointmanager.Endpoints[k]
+	for _, ep := range endpointmanager.Endpoints {
 		go func(ep *endpoint.Endpoint, wg chan<- bool) {
 			scopedLog := log.WithField(logfields.EndpointID, ep.ID)
-			if err := d.allocateIPs(ep); err != nil {
+			ep.Mutex.Lock()
+			if err := d.allocateIPsLocked(ep); err != nil {
```

Comment:
Ideally we should move away from using mutexes outside the structures they protect

Reply:
The intent here is to keep the change in endpoint and the corresponding endpoint.state change in the same critical section so that the state reported by e.g., "cilium endpoint list" is more reliable.

To move all the state changes to the endpoint package would require some more refactoring and exporting daemon's functions that are now private. It would be a good thing to do, but do you think this PR should be held for that, or if this could be addressed in a future PR?